cylc/flow/task_queue.py

```diff
@@ -0,0 +1,197 @@
+        return queues
+
+    def _make_indep(self, in_queues: dict) -> dict:
+        """For classic queues tasks cannot belong to multiple limit groups.
+
+        Latest assigned group takes precedence. So, for example, the "default"
+        group will only contain tasks not assigned to any other queue.
+
+        """
+        queues: Dict[str, Any] = {}
+        seen: Dict[str, str] = {}
+        for qname, qconfig in in_queues.items():
+            queues[qname] = {}
+            queues[qname]["members"] = qconfig["members"]
+            queues[qname]["limit"] = qconfig["limit"]
+            if qname == self.Q_DEFAULT:
+                continue
+            for qmem in qconfig["members"]:
+                # Remove from default queue
+                try:
+                    queues[self.Q_DEFAULT]["members"].remove(qmem)
+                except KeyError:
+                    # Already removed.
+                    pass
+                if qmem in seen:
+                    # Override previous queue assignment.
+                    oldq = seen[qmem]
+                    queues[oldq]["members"].remove(qmem)
+                else:
+                    queues[qname]["members"].add(qmem)
+                seen[qmem] = qname
+        return queues
+
+    def _configure_limiters(self, queues: dict) -> None:
+        """Create a Limiter for each limit group."""
+        for name, config in queues.items():
+            self.limiters[name] = Limiter(config["limit"], config["members"])
+
+    def add(self, itask: TaskProxy) -> None:
+        """Queue a task."""
+        LOG.debug(f"Queue add: {itask.identity}")
+        itask.state.reset(is_queued=True)
+        itask.reset_manual_trigger()
+        self.task_deque.appendleft(itask)
+
+    def release(self, active: Counter[str]) -> List[TaskProxy]:
+        """Release queued tasks."""
+        released: List[TaskProxy] = []
+        rejects: List[TaskProxy] = []
+        while True:
+            try:
+                candidate = self.task_deque.pop()
+            except IndexError:
+                # queue empty
+                break
+            if (candidate.state.is_held or
+                any(limiter.is_limited(candidate, active)
+                    for name, limiter in self.limiters.items())):
+                # Task held, or limited by at least one limit group
+                rejects.append(candidate)
+            else:
+                # Not limited by any groups.
+                candidate.state.reset(TASK_STATUS_PREPARING)
+                candidate.state.reset(is_queued=False)
+                released.append(candidate)
+                active.update({candidate.tdef.name: 1})
+
+        # Re-queue rejected tasks, in the original order.
+        for itask in reversed(rejects):
+            self.task_deque.append(itask)
+        if released:
+            LOG.debug("Queue release:")
+            for r in released:
+                LOG.debug(f"  {r.identity}")
```

Comment:
For logging niceness can write this out as a single message with newlines (the logging is configured to auto-indent following lines), I think this'll do it:

```suggestion
            LOG.debug(
                "Queue release:\n"
                + '\n'.join(f"* {r.identity}" for r in released)
            )
```

It might be good to move this to the calling code so that it doesn't have to be re-implemented in other queue implementations if we go that way.

Reply:
Done.